fish-rust/src/builtins/commandline.rs

```diff
@@ -1,5 +1,501 @@
 
+/// Which part of the comandbuffer are we operating on.
+enum TextScope {
+    String,
+    Job,
+    Process,
+    Token,
+}
+
+/// For text insertion, how should it be done.
+enum AppendMode {
+    // replace current text
+    Replace,
+    // insert at cursor position
+    Insert,
+    // insert at end of current token/command/buffer
+    Append,
+}
+
+/// Replace/append/insert the selection with/at/after the specified string.
+///
+/// \param begin beginning of selection
+/// \param end end of selection
+/// \param insert the string to insert
+/// \param append_mode can be one of REPLACE_MODE, INSERT_MODE or APPEND_MODE, affects the way the
+/// test update is performed
+/// \param buff the original command line buffer
+/// \param cursor_pos the position of the cursor in the command line
+fn replace_part(
+    range: Range<usize>,
+    insert: &wstr,
+    insert_mode: AppendMode,
+    buff: &wstr,
+    cursor_pos: usize,
+) {
+    let mut out_pos = cursor_pos;
+
+    let mut out = buff[..range.start].to_owned();
+
+    match insert_mode {
+        AppendMode::Replace => {
+            out.push_utfstr(insert);
+            out_pos = out.len();
+        }
+        AppendMode::Append => {
+            out.push_utfstr(&buff[range.clone()]);
+            out.push_utfstr(insert);
+        }
+        AppendMode::Insert => {
+            let cursor = cursor_pos - range.start;
+            assert!(range.start <= cursor);
+            out.push_utfstr(&buff[range.start..cursor]);
+            out.push_utfstr(&insert);
+            out.push_utfstr(&buff[cursor..range.end]);
+            out_pos += insert.len();
+        }
+    }
+
+    out.push_utfstr(&buff[range.end..]);
+    commandline_set_buffer(out, Some(out_pos));
+}
+
+/// Output the specified selection.
+///
+/// \param begin start of selection
+/// \param end  end of selection
+/// \param cut_at_cursor whether printing should stop at the surrent cursor position
+/// \param tokenize whether the string should be tokenized, printing one string token on every line
+/// and skipping non-string tokens
+/// \param buffer the original command line buffer
+/// \param cursor_pos the position of the cursor in the command line
+fn write_part(
+    range: Range<usize>,
+    cut_at_cursor: bool,
+    tokenize: bool,
+    buffer: &wstr,
+    cursor_pos: usize,
+    streams: &mut IoStreams,
+) {
+    let pos = cursor_pos - range.start;
+
+    if tokenize {
+        let mut out = WString::new();
+        let buff = &buffer[range];
+        let mut tok = Tokenizer::new(buff, TOK_ACCEPT_UNFINISHED);
+        while let Some(token) = tok.next() {
+            if cut_at_cursor && token.end() >= pos {
+                break;
+            }
+
+            if token.type_ == TokenType::string {
+                let tmp = tok.text_of(&token);
+                let unescaped =
+                    unescape_string(tmp, UnescapeStringStyle::Script(UnescapeFlags::INCOMPLETE))
+                        .unwrap();
```

Comment:
No the C++ panicked as well:

```
-            if (token->type_ == token_type_t::string) {
-                wcstring tmp = *tok->text_of(*token);
-                auto maybe_unescaped = unescape_string(tmp.c_str(), tmp.size(), UNESCAPE_INCOMPLETE,
-                                                       STRING_STYLE_SCRIPT);
-                assert(maybe_unescaped);
-                out.append(*maybe_unescaped);
-                out.push_back(L'\n');
-            }
```

I think the assumption is safe because we have TokenType::string. If it has invalid escapes it's TokenType::error